Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm

```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
         m_cv->invalidateKnownTextureContent(texture);
 }
 
+void GraphicsContextGLCocoa::paintRenderingResultsToCanvas(ImageBuffer& imageBuffer)
+{
+    withDrawingBufferAsPlatformImage([&](PlatformImagePtr&& image) {
+        paintToCanvas(WTFMove(image), imageBuffer.backendSize(), imageBuffer.context());
+    });
+}
+
+void GraphicsContextGLCocoa::paintCompositedResultsToCanvas(ImageBuffer& imageBuffer)
+{
+    withDisplayBufferAsPlatformImage([&](PlatformImagePtr&& image) {
+        paintToCanvas(WTFMove(image), imageBuffer.backendSize(), imageBuffer.context());
+    });
+}
+
+void GraphicsContextGLCocoa::withDrawingBufferAsPlatformImage(std::function<void(PlatformImagePtr&&)> func)
+{
+    if (!makeContextCurrent())
+        return;
+
+    if (!m_displayBufferBacking
+        || m_displayBufferBacking->size() != getInternalFramebufferSize())
+        return;
+
+    prepareTexture();
+
+    if (contextAttributes().premultipliedAlpha) {
+        // Use the IOSurface backed image directly
+
+        GL_ScheduleANGLE();
+
+        // FIXME: Should a displayId be passed here?
+        auto cgContext = m_displayBufferBacking->createPlatformContext();
+        if (cgContext) {
+            // FIXME: Should this be CGIOSurfaceContextCreateImageReference to avoid
+            // creating a copy?
+
+            auto drawingImage = m_displayBufferBacking->createImage(cgContext.get());
+            func(WTFMove(drawingImage));
+
+            // FIXME: Should we be interacting with ImageBufferIOSurfaceBackend
+            // directly? This code is
+            // ImageBufferIOSurfaceBackend::prepareForExternalRead(), but that's
+            CGContextFlush(cgContext.get());
+        }
+    } else {
+        // Since IOSurface-backed images can only support premultiplied alpha,
+        // read the image into a PixelBuffer which can be we used to create a
+        // CGImage.
+        //
+        // FIXME: Can the IOSurface be read into a buffer to avoid the read back via GL?
```

Comment:
So the IOSurface can be accessed via 
```
IOSurface::Locker locker { m_displayBufferBacking->surface() };
```
However:
-  the construction of the pixel buffer and data acquisition is bundled in paintRenderingResultsToPixelBuffer
- IOSurface is in BGRA where as the pixel buffer might need to be in RGBA
- IOSurface has row length != 4*width where as pixel buffer might need to be tightly packed
- Some other aspects

So it might be simpler to not do this in this patch.